crates/red_knot_python_semantic/src/types.rs

```diff
@@ -306,3 +312,359 @@ pub struct IntersectionType<'db> {
+                .fold(
+                    IntersectionTypeBuilder::empty(self.db),
+                    |mut builder, sub| {
+                        builder.intersections.extend(sub.intersections);
+                        builder
+                    },
+                )
+        } else {
+            for inner in &mut self.intersections {
+                inner.add_negative(self.db, ty);
+            }
+            self
+        }
+    }
+
+    #[allow(dead_code)]
+    fn build(self) -> Type<'db> {
+        let mut builder = UnionTypeBuilder::new(self.db);
+        for inner in self.intersections {
+            builder = builder.add(inner.build(self.db));
+        }
+        builder.build()
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug, Clone)]
+struct InnerIntersectionTypeBuilder<'db> {
+    positive: FxOrderSet<Type<'db>>,
+    negative: FxOrderSet<Type<'db>>,
+}
+
+impl<'db> InnerIntersectionTypeBuilder<'db> {
+    fn new() -> Self {
+        Self {
+            positive: FxOrderSet::default(),
+            negative: FxOrderSet::default(),
+        }
+    }
+
+    /// Adds a positive type to this intersection.
+    fn add_positive(&mut self, db: &'db dyn Db, ty: Type<'db>) {
+        match ty {
+            Type::Intersection(inter) => {
+                let pos = inter.positive(db);
+                let neg = inter.negative(db);
+                self.positive.extend(pos.difference(&self.negative));
+                self.negative.extend(neg.difference(&self.positive));
+                self.positive.retain(|elem| !neg.contains(elem));
+                self.negative.retain(|elem| !pos.contains(elem));
+            }
+            _ => {
+                if !self.negative.remove(&ty) {
+                    self.positive.insert(ty);
+                };
+            }
+        }
+    }
+
+    /// Adds a negative type to this intersection.
+    fn add_negative(&mut self, db: &'db dyn Db, ty: Type<'db>) {
+        match ty {
+            Type::Intersection(inter) => {
+                let pos = inter.negative(db);
+                let neg = inter.positive(db);
+                self.positive.extend(pos.difference(&self.negative));
+                self.negative.extend(neg.difference(&self.positive));
+                self.positive.retain(|elem| !neg.contains(elem));
+                self.negative.retain(|elem| !pos.contains(elem));
+            }
+            Type::Never => {}
+            _ => {
+                if !self.positive.remove(&ty) {
+                    self.negative.insert(ty);
+                };
+            }
+        }
+    }
+
+    fn simplify(&mut self) {
+        // TODO this should be generalized based on subtyping, for now we just handle a few cases
+
+        // None intersects with no other type, so we can remove it from negative if there is any
+        // type in positive
+        if !self.positive.is_empty() {
+            self.negative.remove(&Type::None);
+        }
```

Comment:
This isn't correct: `None` intersects with `object`

Reply:
Ah shoot, great catch. There actually isn't any test in this PR depending on this clause, so I just removed it for now.